soc/xtensa/intel_adsp/common/drivers/memory_window/dynamic_slots.c

```diff
@@ -0,0 +1,163 @@
+#endif /* CONFIG_MM_DRV_INTEL_ADSP_MTL_TLB */
+#include <zephyr/arch/xtensa/cache.h>
+
+#if CONFIG_MM_DRV_INTEL_ADSP_MTL_TLB
+static uint8_t tlb[MEMORY_WINDOW_SLOTS_COUNT];
+#endif /* CONFIG_MM_DRV_INTEL_ADSP_MTL_TLB */
+
+uint32_t dynamic_slots_init(const struct device *dev)
+{
+	uint32_t buf_size = MEMORY_WINDOW_SLOT_SIZE * MEMORY_WINDOW_SLOTS_COUNT;
+	const struct mem_win_config *config = dev->config;
+
+	uint32_t *virt_buf = (__sparse_force uint32_t *)
+		arch_xtensa_cached_ptr((void *)(config->mem_base));
+
+	memset(virt_buf, 0, (buf_size+MEMORY_WINDOW_SLOT_SIZE));
+
+	return 0;
+}
+
+uint32_t dynamic_slots_map_slot(uint32_t slot_id, const struct device *dev,
+		uint32_t resource_id, void *virtual_memory_address)
+{
+
+	const struct mem_win_config *config = dev->config;
+
+	uint32_t __sparse_cache *buf = (uint32_t __sparse_cache *)
+		arch_xtensa_cached_ptr((void *)(config->mem_base));
+	uint32_t buf_size =
+		MEMORY_WINDOW_SLOT_SIZE + MEMORY_WINDOW_SLOT_SIZE*MEMORY_WINDOW_SLOTS_COUNT;
+
+	z_xtensa_cache_flush_inv(buf, buf_size);
+
+	struct memory_window_struct *memory_window =
+			(struct memory_window_struct *)(z_soc_uncached_ptr(buf));
+
+	size_t first_free = ARRAY_SIZE(memory_window->slot_descs);
+
+	for (size_t idx = 0; idx < ARRAY_SIZE(memory_window->slot_descs); ++idx) {
+		if (memory_window->slot_descs[idx].vma == (uint32_t)virtual_memory_address &&
+		    memory_window->slot_descs[idx].resource_id == resource_id &&
+		    memory_window->slot_descs[idx].slot_id == slot_id) {
+			return 0;
+		}
+		if (first_free == ARRAY_SIZE(memory_window->slot_descs) &&
+			memory_window->slot_descs[idx].vma == 0) {
+			first_free = idx;
+		}
+	}
+
+	if (first_free == ARRAY_SIZE(memory_window->slot_descs)) {
+		return -ENOBUFS;
+	}
+
+	__ASSERT(memory_window->slot_descs[first_free].slot_id == 0,
+			"Error: Dynamic slot allocation failed. Invalid slot type.");
+	__ASSERT(memory_window->slot_descs[first_free].resource_id == 0,
+			"Error: Dynamic slot allocation failed. Invalid resource id.");
+
+#if CONFIG_MM_DRV_INTEL_ADSP_MTL_TLB
+	uint8_t *slot_address = memory_window->slots[first_free];
+
+	if (virtual_memory_address) {
+		uint32_t err = sys_mm_drv_map_region(
+			(uint32_t __sparse_cache *)arch_xtensa_cached_ptr((void *)slot_address),
+			 0, MEMORY_WINDOW_SLOT_SIZE, SYS_MM_MEM_CACHE_WB);
+
+		if (err != 0) {
+			return err;
+		}
+
+		tlb[first_free] = 1;
+		err = sys_mm_drv_mirror_region(virtual_memory_address,
+					MEMORY_WINDOW_SLOT_SIZE, slot_address, 0);
+		if (err != 0) {
+			return err;
+		}
+		memory_window->slot_descs[first_free].vma = (uint32_t)virtual_memory_address;
+	} else {
+		tlb[first_free] = 0;
+		memory_window->slot_descs[first_free].vma = (uint32_t)slot_address;
+	}
+#endif /* CONFIG_MM_DRV_INTEL_ADSP_MTL_TLB */
+
+	memory_window->slot_descs[first_free].slot_id = slot_id;
+	memory_window->slot_descs[first_free].resource_id = resource_id;
+
+	z_xtensa_cache_flush_inv((buf), MEMORY_WINDOW_SLOT_SIZE);
```

Comment:
now we're effectively asking to overwrite all our configuration, that we've just done bypassing cached with whatever contents caches had before